### PR TITLE
feat: Thingiverse v2 auth/token manager

### PR DESCRIPTION
**Pull request by devonjones**
## Summary

First code PR of the Thingiverse publish/sync epic (`openforge_catalog-4kx`), implementing ticket `openforge_catalog-7dg`.

- **`openforge/thingiverse/auth.py`** — `TokenManager`:
  - `login()` via `POST /v2/auth/login` (the SPA's own flow — no OAuth redirect needed); raises `TwoFactorRequired` on HTTP 202, completed by `login_2fa()` on the same session (the 2FA exchange is cookie-bound)
  - Persists `{access, refresh}` JWTs to `~/.config/openforge/thingiverse_tokens.json` (0600, dir 0700, overridable via `THINGIVERSE_TOKEN_FILE`); **passwords are never stored**
  - `access_token()` checks the JWT `exp` claim (60s leeway, undecodable = expired) and transparently rotates via `POST /v2/auth/refresh`; rejected refresh → `NotLoggedIn` with a clear re-login message
  - `whoami()` (`GET /v2/users/me`) to verify which account the tokens belong to — important given the v1 app token turned out to be scoped to the wrong account
- **`bin/tv_auth`** — click CLI: `login` (prompted credentials, 2FA prompt when needed), `status` (verifies tokens against the live API), `logout`. `-v/-q` verbosity ladder, diagnostics to stderr.
- **`tests/test_thingiverse_auth.py`** — 25 tests; fake session injected at our boundary (per the pack's mocking discipline), fake JWTs with real exp claims, token-file permission assertions, password-never-in-errors assertion.
- **`docs/thingiverse-api-v2.md`** — v2 API reference extracted from the official OpenAPI spec (71 operations, 43 schemas), documenting the auth flows this implements and the read surface later tickets use.

## Test plan

- [x] `pytest tests/test_thingiverse_auth.py` — 25/25 green
- [x] `ruff check .` clean; full suite run (4 pre-existing `test_incremental.py` failures reproduce on clean `test` locally; CI is green there)
- [ ] Live verification: Devon runs `bin/tv_auth login` + `bin/tv_auth status` to confirm a Devon-scoped JWT round-trips against the real API

Beads: `openforge_catalog-7dg` (closed via `bd close` on merge). Unblocks `gn2` (API client) together with the HAR ticket `hnr`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: openforge/thingiverse/auth.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        token_file: Optional[Path] = None,
+        session: Optional[requests.Session] = None,
+    ):
+        if token_file is None:
+            token_file = Path(
+                os.environ.get(TOKEN_FILE_ENV_VAR) or DEFAULT_TOKEN_FILE
+            ).expanduser()
+        self.token_file = Path(token_file)
+        if session is None:
+            session = requests.Session()
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:111]: 🤖 **Claude Code** (resource-leak-reviewer):


[resource-leak-reviewer] [P3] TokenManager.__init__ constructs a requests.Session but the class has no close()/context-manager (__enter__/__exit__) to release it — the session's connection pool is only reclaimed when the object is GC'd. Fine for the current one-shot CLI (bin/tv_auth), but per the task note this TokenManager will be reused by the sync engine iterating many files. If that future code instantiates a new TokenManager per file/iteration rather than holding one long-lived instance, each Session leaks its connection pool the same way per-call session construction in a loop would. Recommend adding a close() (delegating to self.session.close()) and/or context-manager support now, plus a docstring note that callers must construct TokenManager once and reuse it, not per-item in a loop.

File: openforge/thingiverse/auth.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+        # {access, refresh} directly.
+        jwt = body.get("jwt", body)
+        access = jwt.get("access")
+        refresh = jwt.get("refresh")
+        if not access or not refresh:
+            raise ThingiverseAuthError(f"{action} succeeded but response had no tokens")
+        stored = {
+            "access": access,
+            "refresh": refresh,
+            "stored_at": int(time.time()),
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:290]: 🤖 **Claude Code** (dead-code-reviewer):


[dead-code-reviewer] [P3] `stored_at` is written into the persisted token dict but nothing ever reads it back. `_load_tokens()` callers (`access_token()` uses `tokens.get("access")`, `refresh()` uses `tokens.get("refresh")`) never touch `stored_at`, and no other file references it. Grep evidence: `grep -rn stored_at --include='*.py' .` returns only the write site at auth.py:255 (no reads anywhere in the repo). Likely fine as forward-looking metadata (e.g. for a future 'token age' display), but flagging since nothing consumes it today.

File: openforge/thingiverse/auth.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+API_BASE = "https://api.thingiverse.com"
+REQUEST_TIMEOUT = 30
+USER_AGENT = "openforge-catalog-tools"
+
+# Refresh the access token this many seconds before its exp claim.
+EXPIRY_LEEWAY = 60
+
+DEFAULT_TOKEN_FILE = "~/.config/openforge/thingiverse_tokens.json"
+TOKEN_FILE_ENV_VAR = "THINGIVERSE_TOKEN_FILE"
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:38]: 🤖 **Claude Code** (credentials-hygiene-reviewer):


[credentials-hygiene-reviewer] [P3] New env var THINGIVERSE_TOKEN_FILE (overrides where the refresh/access tokens are persisted) isn't named in any doc/README — it only appears in code docstrings and CLI --help. Per project convention (setup shouldn't require reading source), add a one-line mention of this env var name to docs/thingiverse-api-v2.md or a setup doc.

File: openforge/thingiverse/auth.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        if not access or not refresh:
+            raise ThingiverseAuthError(f"{action} succeeded but response had no tokens")
+        stored = {
+            "access": access,
+            "refresh": refresh,
+            "stored_at": int(time.time()),
+        }
+        # The session token accompanies the JWTs on login responses; keep
+        # it in case v1-style endpoints need it later.
+        if "token" in body:
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:294]: 🤖 **Claude Code** (dead-code-reviewer):


[dead-code-reviewer] [P3] `session_token` is stored (from the login response's top-level `token` field) but never read back anywhere in production code — only `tests/test_thingiverse_auth.py:121` asserts it was written. Grep evidence: `grep -rn session_token openforge/thingiverse` and a repo-wide grep show no consumer besides the assignment at auth.py:259-260 and that one test assertion. The comment above it ('keep it in case v1-style endpoints need it later') documents this as intentional future-proofing, so this may be acceptable, but it is dead weight until something reads it — worth confirming that's the intent rather than leftover scaffolding.

File: openforge/thingiverse/auth.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        # JWT segments are base64url without padding
+        payload_b64 += "=" * (-len(payload_b64) % 4)
+        payload = json.loads(base64.urlsafe_b64decode(payload_b64))
+        exp = payload.get("exp")
+        return int(exp) if exp is not None else None
+    except (IndexError, ValueError, binascii.Error, json.JSONDecodeError):
+        return None
+
+
+def _token_expired(token: str, leeway: int = EXPIRY_LEEWAY) -> bool:
```
Comment (devonjones) [inline openforge/thingiverse/auth.py]: 🤖 **Claude Code** (dead-code-reviewer):


[dead-code-reviewer] [P3] `_token_expired`'s `leeway` parameter is never invoked with a non-default value anywhere in the repo. Grep evidence: `grep -rn '_token_expired(' --include='*.py' .` shows the single production call (auth.py:166: `_token_expired(access)`) and four test calls (tests/test_thingiverse_auth.py:99,102,106,109), none of which pass `leeway`. Not necessarily worth removing (useful seam for tests to exercise custom leeway), but currently unused surface — every caller relies on `EXPIRY_LEEWAY`.

File: tests/test_thingiverse_auth.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    return f"{seg({'alg': 'none'})}.{seg(payload)}.fakesig"
+
+
+class FakeResponse:
+    def __init__(self, status_code, body=None):
+        self.status_code = status_code
+        self._body = body if body is not None else {}
+
+    def json(self):
+        if isinstance(self._body, Exception):
```
Comment (devonjones) [inline tests/test_thingiverse_auth.py:42]: 🤖 **Claude Code** (dead-code-reviewer):


[dead-code-reviewer] [P3] `FakeResponse.json()`'s exception-simulation branch (`if isinstance(self._body, Exception): raise self._body`) is unused test-helper dead code. Grep evidence: `grep -n "FakeResponse(" tests/test_thingiverse_auth.py` shows every call site passes a dict body (e.g. `{"error": "..."}`, `{}`); none constructs `FakeResponse` with an Exception. This branch appears intended to exercise `_handle_token_response`'s `except (ValueError, AttributeError)` fallback around `resp.json()` (auth.py:235-238), but no test currently covers that path — either add a test that passes an Exception body, or drop the branch.

File: openforge/thingiverse/auth.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    """
+
+    def __init__(
+        self,
+        token_file: Optional[Path] = None,
+        session: Optional[requests.Session] = None,
+    ):
+        if token_file is None:
+            token_file = Path(
+                os.environ.get(TOKEN_FILE_ENV_VAR) or DEFAULT_TOKEN_FILE
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:107]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P1] TokenManager.__init__'s token_file default-resolution branch (os.environ.get(TOKEN_FILE_ENV_VAR) or DEFAULT_TOKEN_FILE) is never exercised — every test in tests/test_thingiverse_auth.py passes token_file explicitly via the token_file fixture. This is real precedence logic (THINGIVERSE_TOKEN_FILE env var vs default path), not glue. Coverage run confirms line 103 (the assignment) is uncovered. Add two tests: (1) env var set -> TokenManager(token_file=None) resolves to that path, (2) env var unset -> resolves to DEFAULT_TOKEN_FILE (expanduser'd).

File: openforge/thingiverse/auth.py [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+        Returns:
+            The new access token (the rotated refresh token is persisted)
+
+        Raises:
+            NotLoggedIn: If no refresh token is stored or it's rejected
+        """
+        tokens = self._load_tokens()
+        refresh_token = tokens.get("refresh")
+        if not refresh_token:
+            raise NotLoggedIn("no refresh token stored; log in first")
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:189]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P1] refresh()'s 'no refresh token stored' branch (tokens.get("refresh") falsy -> raise NotLoggedIn) is uncovered. Coverage run confirms line 186 is uncovered. Existing tests don't reach it: test_no_token_file_raises_not_logged_in fails earlier in _load_tokens (no file at all), and test_rejected_refresh_raises_not_logged_in always has a refresh token present in the stored tokens. Need a test where the token file exists with valid JSON but no 'refresh' key (e.g. only {"access": ...}), calling access_token() or refresh() directly and asserting NotLoggedIn with message matching 'no refresh token stored'.

File: openforge/thingiverse/auth.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+    # -- internals ------------------------------------------------------
+
+    def _handle_token_response(self, resp: requests.Response, action: str) -> Dict:
+        """Validate a token-bearing response and persist its tokens."""
+        if resp.status_code != 200:
+            # Response bodies for auth errors are {"error": "..."} per the
+            # API doc; include the message but never echo tokens.
+            detail = ""
+            try:
+                detail = resp.json().get("error", "")
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:269]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P2] _handle_token_response's except (ValueError, AttributeError) fallback (lines 236-238), hit when resp.json() itself raises while extracting the error detail from a non-200 response, is uncovered per the coverage run. All existing non-200 FakeResponse fixtures (401, 202, 500) return valid dict bodies so .json() always succeeds. Add a test where the error-response body raises on .json() (FakeResponse already supports this via body=<Exception>) and assert the resulting message is 'HTTP <code>' with no bracketed detail suffix.

File: openforge/thingiverse/auth.py (hunk 1 above)
Comment (devonjones) [inline openforge/thingiverse/auth.py:111]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P3] TokenManager's default-session construction (session=None -> requests.Session() with User-Agent header, lines 108-109) is never exercised — every test injects FakeSession via the session fixture, so this branch shows 0% coverage. Low severity since it's mostly object construction, but a quick test (TokenManager(token_file=tmp_path/...) with no session arg, asserting isinstance(manager.session, requests.Session) and the User-Agent header) would close the gap cheaply.

File: bin/tv_auth [hunk 10]
```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
```
Comment (devonjones) [inline bin/tv_auth:1]: 🤖 **Claude Code** (test-coverage-reviewer):


[test-coverage-reviewer] [P1] bin/tv_auth has zero tests (no unit test, and no integration test either — integration_tests/ has nothing for it). This file is more than glue: _configure_logging() computes a log level from quiet/verbose with real branching (lines 48-53); login() has a nested try/except that catches TwoFactorRequired and re-prompts for a 2FA code before retrying (lines 73-79); status() branches on is_logged_in() and distinguishes NotLoggedIn from ThingiverseAuthError with different exit codes (lines 91-101). Per the review spec, the integration-test exemption only covers glue with no branching logic — this doesn't qualify, and no integration test exists regardless. click.testing.CliRunner would make these cheap to test: login without 2FA, login with the TwoFactorRequired->prompt->login_2fa retry path, status when not logged in (exit 1), status when tokens are stored but unusable (NotLoggedIn, exit 1, distinct message), and status success.

**Review by devonjones — COMMENTED** [review]
## Code Review (Claude)

This review focuses on auth-logic correctness, API contract fidelity against
`docs/thingiverse-api-v2.md`, and house style. Complexity, error-handling
depth, test coverage, resource leaks, dead code, logging, and credentials
hygiene are covered by other reviewers and are intentionally not duplicated
here.

### Critical Issues

**1. `bin/tv_auth` is committed non-executable, breaking the documented usage.**

```
$ git ls-files -s bin/tv_auth bin/db_update
100644 ... bin/tv_auth   <- no exec bit
100755 ... bin/db_update <- exec bit (every other bin/ tool)
```

Every other script in `bin/` is mode `100755`. `tv_auth` is `100644`, so
`bin/tv_auth login` (as documented in the PR body and the module docstring)
will fail with "Permission denied" on a fresh checkout — it only works via
`python bin/tv_auth ...`. This directly violates the Unix-tooling convention
in `openforge/CLAUDE.md` ("Command-line tools go in `bin/` directory... use
standard Unix conventions"). Fix: `chmod +x bin/tv_auth && git add
--chmod=+x bin/tv_auth` (or `git update-index --chmod=+x bin/tv_auth`) before
merge.

**2. `_jwt_exp` crashes (uncaught `AttributeError`) on a structurally-valid
JWT whose payload segment decodes to non-object JSON**, contradicting its own
contract and the "undecodable = expired" fail-safe design.

`openforge/thingiverse/auth.py:52-70`:
```python
payload = json.loads(base64.urlsafe_b64decode(payload_b64))
exp = payload.get("exp")
```
only `IndexError, ValueError, binascii.Error, json.JSONDecodeError` are
caught. If the payload segment is valid JSON but not a dict (e.g. a JSON
array, string, or number — plausible for a hand-edited/corrupted token file,
or an API/format change), `payload.get` raises `AttributeError`, which is
**not** caught and propagates out of `_token_expired()` → `access_token()`,
crashing the whole auth path instead of the documented "can't be decoded ->
None -> treated as expired" behavior:

```python
>>> _jwt_exp(f"{seg({'alg':'none'})}.{seg([1,2,3])}.sig")
AttributeError: 'list' object has no attribute 'get'
```
Since the token file (`~/.config/openforge/thingiverse_tokens.json`) is
user-editable, untrusted, unsigned input, this fail-safe gap is a real risk,
not just theoretical. Fix: add `AttributeError` to the except tuple (or
`isinstance(payload, dict)` check before `.get`).

### Medium Issues

**3. Refresh-token rotation is neither atomic nor crash-safe, and the file
has no cross-process locking — two concerns that compound.**

`_store_tokens()` (`auth.py:265-272`) opens the target file directly with
`O_TRUNC` and writes in place:
```python
fd = os.open(self.token_file, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
```
If the process dies (OOM-kill, `Ctrl-C`, laptop sleep) between the server
accepting the refresh call (which very likely invalidates/rotates the old
refresh token server-side) and this write completing, the on-disk file is
left empty or truncated, and the *previous* refresh token — now already
consumed — is what a subsequent run would try to reuse. The result is a full
forced re-login, and depending on timing, a JSON-corrupt file rather than a
clean rollback. There's no temp-file + `os.replace()` for atomicity, and no
`fcntl.flock`/lockfile serializing concurrent access — so two concurrent
`tv_auth`/future-sync-tool invocations reading the same refresh token will
race: whichever POSTs to `/v2/auth/refresh` second gets a `401` on an
already-rotated token and is forced into `NotLoggedIn`, aborting whatever
batch operation triggered it. Given this is meant to back an unattended sync
tool (per the PR's stated epic), both the non-atomic write and the missing
lock are worth addressing before other tools build on `TokenManager`
(write-to-temp-then-rename is a small, cheap fix; locking can be a simple
advisory `flock` around `refresh()`/`_store_tokens()`).

**4. The 2FA cookie-binding assumption is not established anywhere in
`docs/thingiverse-api-v2.md`, and is untested against a live 2FA account.**

`login_2fa()`'s docstring and the module docstring assert "the 2FA exchange
is tied to the login attempt's session cookies," but nothing in the API doc
(or the OpenAPI-derived `Authenticate2FARequest {code}` schema) documents
cookie semantics — the doc is silent on transport beyond the JWT Bearer
scheme. The assumption is *reasonable by elimination* (the 2FA endpoint's
request body carries only `code`, so a cookie is the only remaining channel
for the server to know which login attempt the code belongs to), but it's
inference, not spec. The PR's test plan explicitly leaves live verification
unchecked, and doesn't call out that the 2FA path specifically needs
verification against an account that actually has 2FA enabled. If the
assumption is wrong, the failure mode is an opaque `ThingiverseAuthError:
2FA login failed: HTTP 401` with no hint that the real problem is "must
reuse the same `TokenManager`/session as the initial `login()` call."
Recommend: (a) add a line to `docs/thingiverse-api-v2.md`'s auth section
noting this is an inferred/observed behavior, not from the spec, and (b)
special-case a 401/403 from `login_2fa()` with a clearer message.

**5. `logout()` only deletes the local token file; it never calls the API's
own logout endpoint.** `docs/thingiverse-api-v2.md` documents `GET
/v2/auth/logout` ("Logout, clear login cookies"). Local-only deletion means
the refresh token is never revoked server-side — it remains valid until it
naturally expires, so "logout" doesn't actually invalidate the session,
just the local copy. That may be an intentional/acceptable scope decision for
a CLI dotfile tool, but it's worth a one-line docstring caveat so future
callers (e.g. an eventual "revoke all sessions" feature) don't assume
`logout()` does more than it does.

### Minor Issues

- `auth.py:247` — `jwt = body.get("jwt", body)` correctly distinguishes
  `AuthTokensResponse` (has `jwt` key) from `JwtTokenResponse` (returned
  directly by refresh), matching the doc's two shapes. But if a response
  ever included an explicit `"jwt": null` or `"jwt": {}`, `.get("jwt", body)`
  would return that falsy value rather than falling back to `body`, and
  `jwt.get("access")` would raise `AttributeError` on `None`. Low
  probability given the doc's schemas, but same class of bug as #2 — an
  `isinstance` guard would make both spots equally robust.
- `bin/tv_auth`'s `status` command writes its failure states
  ("Not logged in.", "Tokens stored but unusable: ...") via `click.echo(...)`
  to stdout, then `sys.exit(1)` — per the Unix-convention guidance in
  `openforge/CLAUDE.md` ("stderr for errors"), these should probably be
  `click.echo(..., err=True)` so a non-zero exit code is paired with stderr
  output, consistent with how `login`/`logout` already route failures
  through `click.ClickException` (which Click prints to stderr automatically).
- The refresh endpoint's `401 -> NotLoggedIn` special-case (`auth.py:192`)
  assumes `/v2/auth/refresh` behaves like an authenticated route for its
  failure status code; the doc's only stated 401 semantics are for
  `Authorization: Bearer` routes in general, and `/v2/auth/refresh` takes
  the refresh token as a body param, not a Bearer header. Not necessarily
  wrong, just unconfirmed by the doc — worth a quick manual check when
  Devon does the live verification pass.

### Positive Observations

- Login/2FA/refresh request payload key casing matches
  `docs/thingiverse-api-v2.md` exactly, including the (Thingiverse-side)
  inconsistency between `usernameOrEmail` (camelCase, login) and
  `refresh_token` (snake_case, refresh) — the code doesn't try to
  "normalize" this away, which is the right call.
- `_jwt_exp`'s base64url padding math (`"=" * (-len(payload_b64) % 4)`) is
  correct for all four padding cases and is exercised well by the
  `TestJwtHelpers` tests.
- `TokenManager` accepting an injectable `session` is exactly the right
  seam for testing an HTTP-boundary class without a mocking framework, and
  matches how `SessionService`/`IncrementalScanner` do similar things
  elsewhere in the codebase — a class here is consistent with existing
  precedent for stateful resources (session + token file path), despite the
  general procedural-style preference in `openforge/CLAUDE.md`.
- Token file permissions (`0600` file, `0700` dir) are set correctly and
  proactively via `os.open`/`os.chmod` rather than relying on umask.
- `ruff check` and `ruff format --diff` are clean on all four changed
  Python files; `pytest tests/test_thingiverse_auth.py` is 25/25 green.

### Recommendations Summary

**Must fix before merge:**
- Set the executable bit on `bin/tv_auth` (Critical #1).
- Guard `_jwt_exp` against non-dict JSON payloads so malformed/corrupt
  tokens degrade to "expired" instead of crashing (Critical #2).

**Should fix:**
- Make `_store_tokens` atomic (write-temp + `os.replace`) and add a simple
  lock around refresh/persist to avoid the concurrent-invocation race
  (Medium #3).
- Document the 2FA cookie-binding assumption as inferred (not spec-backed)
  in `docs/thingiverse-api-v2.md`, and give `login_2fa()` failures a clearer
  error message (Medium #4).
- Add a one-line docstring caveat that `logout()` is local-only and doesn't
  revoke the refresh token server-side (Medium #5).

**Nice to have:**
- Guard the `jwt.get(...)` call similarly to #2 for the `{"jwt": null}` edge
  case (Minor).
- Route `status`'s failure-path `click.echo` calls to stderr (Minor).
- Confirm refresh's 401 semantics during the live-verification pass (Minor).

### Overall Assessment

Solid, well-tested first cut of the auth layer with good API-contract
fidelity to `docs/thingiverse-api-v2.md`; the two must-fix items are cheap
(a chmod and an except clause) but both are the kind of thing that bites
silently in production (a locked-out sync job, a script nobody can execute
directly), so worth closing before this becomes the foundation other
Thingiverse tickets build on.

File: openforge/thingiverse/auth.py (hunk 1 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:111]: Fixed - added close() and context-manager support to TokenManager with a docstring directing long-running callers (the sync engine) to reuse one instance per process; test_context_manager_closes_session covers it.

File: openforge/thingiverse/auth.py (hunk 2 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:290]: Won't fix - stored_at is deliberately human-readable metadata for inspecting the token file (when did I last log in), not program state; code intentionally never reads it.

File: openforge/thingiverse/auth.py (hunk 3 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:38]: Fixed - THINGIVERSE_TOKEN_FILE documented in docs/thingiverse-api-v2.md under 'OpenForge tooling environment variables' (name only, no values).

File: openforge/thingiverse/auth.py (hunk 4 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:294]: Won't fix - documented future-proofing: the session token accompanies the JWTs and the epic may need it for v1-style endpoints (the file-download surface isn't settled until the HAR contract lands, openforge_catalog-hnr). Cheap to keep, annoying to re-plumb.

File: openforge/thingiverse/auth.py (hunk 5 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py]: Fixed - removed the leeway parameter; _token_expired now uses EXPIRY_LEEWAY directly.

File: tests/test_thingiverse_auth.py (hunk 6 above)
Comment (devonjones) [reply tests/test_thingiverse_auth.py:42]: Fixed - the branch is now exercised: test_non_json_error_body_still_raises_with_status queues FakeResponse(500, ValueError(...)) to cover _handle_token_response's error-detail-extraction fallback.

File: openforge/thingiverse/auth.py (hunk 7 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:107]: Fixed - TestConstruction adds tests for THINGIVERSE_TOKEN_FILE resolution and the default-path fallback (monkeypatched env). Module coverage now 100%.

File: openforge/thingiverse/auth.py (hunk 8 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:189]: Fixed - test_refresh_with_no_refresh_key_raises_not_logged_in writes a token file lacking the refresh key and asserts NotLoggedIn.

File: openforge/thingiverse/auth.py (hunk 9 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:269]: Fixed - test_non_json_error_body_still_raises_with_status covers the except (ValueError, AttributeError) fallback; the HTTP status still surfaces in the raised error.

File: openforge/thingiverse/auth.py (hunk 1 above)
Comment (devonjones) [reply openforge/thingiverse/auth.py:111]: Fixed - test_default_session_is_requests_session_with_user_agent exercises the session=None path (real requests.Session constructed, User-Agent set, no network).

File: bin/tv_auth (hunk 10 above)
Comment (devonjones) [reply bin/tv_auth:1]: Fixed - CLI logic extracted to openforge/thingiverse/cli.py (bin/tv_auth is now a 3-line shim) and tests/test_thingiverse_cli.py adds 13 CliRunner tests: verbosity ladder math (parametrized), 2FA prompt flow, bad-credential failure, status exit codes, logout. cli.py at 100% coverage.

File: openforge/thingiverse/auth.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+        local token file.
+        """
+        if self.is_logged_in():
+            try:
+                self.session.get(
+                    f"{API_BASE}/v2/auth/logout",
+                    headers=self.auth_header(),
+                    timeout=REQUEST_TIMEOUT,
+                )
+            except (ThingiverseAuthError, requests.RequestException) as e:
```
Comment (devonjones) [inline openforge/thingiverse/auth.py:232]: 🤖 **Claude Code** (test-coverage-reviewer):


[P3] logout()'s new `except (ThingiverseAuthError, requests.RequestException)` (line 232) is only exercised via the requests.RequestException branch — `test_logout_removes_file_when_revocation_fails` uses an ExplodingSession raising `requests.ConnectionError`. The ThingiverseAuthError branch (e.g. `auth_header()` -> `access_token()` -> `refresh()` raising NotLoggedIn when the stored access token is expired and no refresh token is present, all before any network call) is never hit by a test. Coverage tooling reports the line as covered because it's a single except statement (no --cov-branch), but the two exception types represent genuinely different revocation-failure causes. Low severity since both branches converge on identical behavior (log a warning, then proceed to unlink the token file) — a quick addition: write a token file with only an expired access token and no refresh key, then call logout() and assert it still removes the file without raising.

Comment (devonjones) [reply openforge/thingiverse/auth.py:232]: Fixed - test_logout_removes_file_when_tokens_unusable covers the ThingiverseAuthError branch: expired access + missing refresh key raises NotLoggedIn from auth_header before any network call; asserts the file is still removed and no revoke call is attempted. 53 tests passing.
